vms/platformvm/validators_set_property_test.go

```diff
@@ -0,0 +1,856 @@
+		uint64(data.StartTime.Unix()),
+		uint64(data.EndTime.Unix()),
+		data.NodeID,
+		addr,
+		reward.PercentDenominator,
+		[]*secp256k1.PrivateKey{keys[0], keys[1]},
+		addr,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create AddValidatorTx, %w", err)
+	}
+	return internalAddValidator(vm, signedTx)
+}
+
+func internalAddValidator(vm *VM, signedTx *txs.Tx) (*state.Staker, error) {
+	stakerTx := signedTx.Unsigned.(txs.StakerTx)
+	if err := vm.Builder.AddUnverifiedTx(signedTx); err != nil {
+		return nil, fmt.Errorf("could not add tx to mempool, %s", err.Error())
+	}
+
+	blk, err := vm.Builder.BuildBlock(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("failed building block, %s", err.Error())
+	}
+	if err := blk.Verify(context.Background()); err != nil {
+		return nil, fmt.Errorf("failed verifying block, %s", err.Error())
+	}
+	if err := blk.Accept(context.Background()); err != nil {
+		return nil, fmt.Errorf("failed accepting block, %s", err.Error())
+	}
+	if err := vm.SetPreference(context.Background(), vm.manager.LastAccepted()); err != nil {
+                return nil, fmt.Errorf("failed getting last accepted block, %s", err.Error())
+	}
+
+	// move time ahead, promoting the validator to current
+	currentTime := stakerTx.StartTime()
+	vm.clock.Set(currentTime)
+	vm.state.SetTimestamp(currentTime)
+
+	blk, err = vm.Builder.BuildBlock(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("failed building block, %s", err.Error())
+	}
+	if err := blk.Verify(context.Background()); err != nil {
+		return nil, fmt.Errorf("failed verifying block, %s", err.Error())
+	}
+	if err := blk.Accept(context.Background()); err != nil {
+                return fmt.Errorf("failed accepting block, %s", err.Error())
+	}
+	if err := vm.SetPreference(context.Background(), vm.manager.LastAccepted()); err != nil {
+		return fmt.Errorf("failed getting last accepted block, %s", err.Error())
+	}
+
+	return vm.state.GetCurrentValidator(stakerTx.SubnetID(), stakerTx.NodeID())
+}
+
+func terminateSubnetValidator(vm *VM, validator *state.Staker) error {
+	currentTime := validator.EndTime
+	vm.clock.Set(currentTime)
+	vm.state.SetTimestamp(currentTime)
+
+	blk, err := vm.Builder.BuildBlock(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed building block, %s", err.Error())
+	}
+	if err := blk.Verify(context.Background()); err != nil {
+		return fmt.Errorf("failed verifying block, %s", err.Error())
+	}
+	if err := blk.Accept(context.Background()); err != nil {
+		return fmt.Errorf("failed accepting block, %s", err.Error())
+	}
+	if err := vm.SetPreference(context.Background(), vm.manager.LastAccepted()); err != nil {
+		return fmt.Errorf("failed getting last accepted block, %s", err.Error())
+	}
+
+	// _, err = vm.state.GetCurrentValidator(constants.PrimaryNetworkID, nodeID)
+	// require.ErrorIs(err, database.ErrNotFound)
+	return nil
+}
+
+func terminatePrimaryValidator(vm *VM, validator *state.Staker) error {
```

Comment:
Isn't `terminatePrimaryValidator` the same as `terminateSubnetValidator`? I think if we use staker.SubnetID instead of constants.PrimaryNewtorkID we can reuse this function for both cases right?

Reply:
well the test subnet is a permissioned one (no reward for its validator), so subnet validator is terminated by a standard block just advancing chain time, while primary validator require a proposal block + commit bloc acceptance. Reuse requires an if/else in fuction implementation.
I reckon there is some code duplication here, but having two separate methods makes easier to reproduce the faulty, simplified scenario, since it's very obvious what method we should pick.